src/workerd/api/actor.h

```diff
@@ -13,27 +13,95 @@
 #include <capnp/compat/http-over-capnp.h>
 #include <workerd/api/http.h>
 #include <workerd/jsg/jsg.h>
+#include <workerd/jsg/function.h>
 
 namespace workerd {
   template <typename T> class IoOwn;
 }
 
 namespace workerd::api {
 
+enum IsInHouse {
+  YES,
+  NO,
+};
+
+// This class is the client RPC stub for Durable Objects. It intercepts dynamic method calls, then
+// sends the method name and arguments to the remote Durable Object, returning the response it
+// receives.
+//
+// We don't want name interception on Fetcher itself -- at least not yet -- since not all Fetcher
+// implementations need it. Name interception is off by default; to turn it on, set the JsRpcStub
+// compatibility flag.
+class DurableObjectRpcStub : public Fetcher, public jsg::NamedIntercept {
```

Comment:
Can we put this functionality into `Fetcher` itself, rather than a DO-specific subclass? Eventually we want all fetchers to support RPC, so why not go ahead and put the code there now?

Reply:
@jasnell @bcaimano since this came up in the other PR too. We've had some discussions in DO standup about other fetchers being able to use this too but IIRC we wanted to do DOs first, then transition to others later. Any opinions here?

Reply:
My concern here would be that making a common object support named intercept could make it a bit difficult to ensure backwards compat if we need to introduce other non-dynamic apis into the same class later (since those could end up conflicting with dynamic apis). `Fetcher` is one of those that I would be concerned about here. At the same time, I 100% get what @kentonv is suggesting and why. Not sure I have a great counter suggestion tho.

Reply:
I'd like to get this merged (followed by the server side), do a bit of testing in DOG (hopefully this week) so we can start looking for bugs, then decide on moving into the `Fetcher` either before or after releasing. Is this alright?

Reply:
My concern here is that introducing a new type into the hierarchy is an application-visible change and it's possible that it proves to have subtle differences compared to baking things into `Fetcher`, making it hard to change later. I do not think we should roll out this application-visible change for testing purposes if we are not confident it's what we want.

Moreover, I am worried that if we focus too hard on the Durable Object use case we might accidentally create something that only really works for Durable Objects. We know that we do not want this to be DO-specific so I'd like to make sure we don't paint ourselves into that corner.

The `Fetcher` type is badly-named, it is really the catch-all type for a remote endpoint that supports various comment request types -- currently HTTP and sockets. It seems natural to extend this for RPC.

That said I could maybe be convinced that `Fetcher` should remain the type used for non-Workers remote endpoints, and we should really have a subclass called something like `WorkerRpc` which represents a remote Worker/DO. This type would extend `Fetcher` and add protocols that only work within the Workers network, namely JS RPC.

I suppose that is essentially what `DurableObjectRpcStub` is. But I think we should not name this in a way that is specific to durable objects, and we should put it in a separate file from `actor.h/c++`, to show that it's ready to serve as the type for service bindings as well.

> My concern here would be that making a common object support named intercept could make it a bit difficult to ensure backwards compat if we need to introduce other non-dynamic apis into the same class later

This is a risk we've already discussed and decided to take, in order to get the usability benefits of RPC. If we need to introduce a new method later, we'll have to use a compat flag. This is true regardless of whether we extend `Fetcher` or use some derived type here.

> I'd like to get this merged (followed by the server side), do a bit of testing in DOG (hopefully this week) so we can start looking for bugs, then decide on moving into the Fetcher either before or after releasing. Is this alright?

Why does this need to be in DOG to test? It seems like this should be fully testable locally. I'd rather not push out application-visible changes that we aren't sure about yet.

Reply:
> I suppose that is essentially what DurableObjectRpcStub is. But I think we should not name this in a way that is specific to durable objects, and we should put it in a separate file from actor.h/c++, to show that it's ready to serve as the type for service bindings as well.

Yeah, I'd like this as well, honestly.

Reply:
I like the idea of having `WorkerRpc` outside of `actor.h/c++`, and having it extend `Fetcher`. I also wonder if we might as well move the `JsRpcService/Target` code into the same files in the next PR (Kenton [mentioned](https://github.com/cloudflare/workerd/pull/1028#discussion_r1379316975) combining the JsRpc stuff anyways).

I'll put something up soon, have been dealing with some technical issues (OS can't build main, can't upgrade OS because on-call. Have rebased to an older commit now).